### PR TITLE
api: async mode in ext_proc

### PR DESCRIPTION
**Pull request by tyxia**
API changes   to ext_proc async mode. Implementation will be followed once consensus is reached on the api changes/design.

 **Highlighted Changes**:

1. For the body, async mode is restricted to [STREAMED](https://github.com/envoyproxy/envoy/blob/9a47bc9eff9c3283fbfaed9144313b7bda366aac/api/envoy/extensions/filters/http/ext_proc/v3/processing_mode.proto#L58-L60) processing mode only. Alternatively, we can add a new mode (e.g., ASYNC_STREAMED mode). I don't feel strong either way, please let me know your thoughts.
2. Specify that ext_proc server' response is not expected in this mode and will be ignored .
3. Add warning section to state that flow control is not ready yet.

### CODE REVIEW

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 1]
```diff
@@ -115,12 +115,25 @@ message ExternalProcessor {
...
   ProcessingMode processing_mode = 3;
 
-  // [#not-implemented-hide:]
   // If true, send each part of the HTTP request or response specified by ProcessingMode
   // asynchronously -- in other words, send the message on the gRPC stream and then continue
-  // filter processing. If false, which is the default, suspend filter execution after
+  // filter processing. In this mode:
+  // 1) Only STREAMED body processing mode is supported and any other body processing modes
+  // will be ignored for asynchronous processing.
+  // 2) ext_proc server is not expected to send back processing response. Any response from
```
Comment (yanavlasov) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Is this going to be the final state? This will limit the usefulness of the feature.

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Sorry, which part do you think will limit the `usefulness of the feature`? (1) Only STREAMED mode or (2) no response?

Comment (yanavlasov) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: No response

Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: It is good topic to discuss.

In terms of Envoy's general capabilities: IMHO Envoy can not fully support async_mode + external processor response together.  As defined in the existing comment, async mode means no pause on filter chain iteration. Header event for example, if header is propagated to remaining filters in the chain, the external processor's response/mutation is too late.

In terms of our current use case of async mode: we want to use it for observability purpose : Send and Go. We just forward Envoy data to external processor to observe and no need for mutation/response.
 
In short, I think async mode can not fully support response. But if it is the concern/blocker here, alternatively I think we can introduce a new mode `observability_mode` to achieve our purpose without touching async mode.

Comment (yanavlasov) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: The "async" mode was never defined well. It is possible to turn it into purely "observability" mode where client request is immutable. This is fine, but I would go with adding a new member that clearly describes the mode - calling observability mode "async" is confusing.

To me 'async' implied that ext_proc filter does not wait for response from ext_proc server before sending the next ext_proc request to it. I.e. it can send body without first waiting for ext_proc server to respond to the headers request. It would need to pause the filter chain if header mutation is expected (I do not think it needs to pause filter chain for body).

In any case if you are implementing the read-only observability mode, I would use a new config for it and not reuse the `async` config option.

Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Thank you for detailed response!
I think there are two options here:
(1) Add a new config field `observability_mode `.
(2) Reuse the existing config filed `async_mode`. We need to clearly define it (probably slightly change people's interpretation).

Either works for me. @htuch what is your thoughts on this? Thanks

Comment (htuch) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Another approach here is to drop the coarse grained `async_mode`, which is not implemented or used, and provide finer-grained control in `ProcessingMode`. For example, we could have a `HeaderSendMode` that was `SEND_NO_WAIT`. @gbrail for thoughts as well.

Comment (gbrail) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Yes I see now long after writing it why the description of "async mode" confuses people!

The intention was indeed for it to be a "fire and forget" mode for observability. In async_mode, Envoy would not expect any responses from the processor, other than closing the stream if it was no longer interested in the conversation.

Perhaps it should have been called "oneway" instead.

Normally, ext_proc works like this:

1. Send request headers to processor
2. Wait for request headers response
3. Apply response and continue filter chain
4. Send body chunks to processor as they are delivered by the filter chain
5. Apply body chunk responses from the processes in order to the chunks that were sent out
6. Depending on flow control status, possibly invoke flow control to prevent the set of chunks sent in step 4 from growing too much, thus slowing down step 4

In async_mode, ext_proc would instead work like this:

1. Send request headers to processor with "async_mode" flag set to true as shown in https://github.com/envoyproxy/data-plane-api/blob/fe9be030c0b05e02904156b417bfcc1f235936fe/envoy/service/ext_proc/v3/external_processor.proto#L62]
2. Immediately continue filter chain
3. Send each body chunk to the processor with async_mode set to true

In other words -- today, Envoy expects the processor to respond to each stream message with a matching reply. In async_mode, Envoy would expect the processor never to respond, and might error or just ignore any messages that were sent back. This would make it useful for monitoring as it would support monitoring without adding latency.

The reason I never implemented that (other than that I ran out of time) was because I couldn't figure out flow control for streamed messages. If we send each message body chunk to the processor when it arrives, with no flow control, then if the sender is much faster than the processor, what will break? I don't understand flow control in all the parts of Envoy well enough to know how to address that.

Finally, on the idea that async mode meant that Envoy would still expect a response from the processor, but would send body chunks first -- that could be theoretically possible, as long as we don't expect the processor to be able to modify the headers. Once Envoy has moved into streaming the body, the headers have already been sent along the filter chain and may have already been transmitted on the socket. To give the processor a chance to modify the headers, Envoy would have to ignore the headers responses.

That would imply a different meaning to "async_mode" -- not that we don't expect a reply from the processor, but that any replies will be ignored. That fixes the control plane problem, and removes the need for the "async_mode" flag on each stream message, which was supposed to tell the processor not to send a reply to a particular message.

Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Thanks very much for review and info!
> In async_mode, Envoy would expect the processor never to respond, and might error or just ignore any messages that were sent back. This would make it useful for monitoring as it would support monitoring without adding latency.

Based on @gbrail's response above, the original idea of async_mode is no-response observability mode, which this PR is in accordance with and this PR clearly documents that.

@yanavlasov  WDYT about proceeding with this PR as it is and add a new `async_with_response` mode if it is needed in the future? If the separate `observability_mode ` is still preferred, PLMK.

Comment (yanavlasov) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: Leaving it as is will not reduce confusion. There is also danger of implementing poorly defined configuration elements that were not implemented for a long time - there is nothing preventing operators from actually setting these fields in the config by mistake (as we have recently learned), making deployment problematic. We should really create a new configuration element.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:123]: I have created https://github.com/envoyproxy/envoy/pull/32696 for adding new configuration field. PTAL, Thanks!

(i kept this PR open in case reviewers on this thread want to discuss further)

**Comment by tyxia** [comment]
close in favor of https://github.com/envoyproxy/envoy/pull/32696
